src/view/switch.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2022 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+
+use crate::view::ViewMarker;
+use crate::{view::Id, widget::ChangeFlags, MessageResult};
+
+use super::{Cx, View};
+
+pub struct Switch {
+    is_on: bool,
+}
+
+pub fn switch(is_on: bool) -> Switch {
+    Switch::new(is_on)
+}
+
+impl Switch {
+    pub fn new(is_on: bool) -> Self {
+        Switch { is_on }
+    }
+}
+
+impl ViewMarker for Switch {}
+
+impl<A> View<bool, A> for Switch {
+    type State = ();
+
+    type Element = crate::widget::Switch;
+
+    fn build(&self, cx: &mut Cx) -> (crate::view::Id, Self::State, Self::Element) {
+        let (id, element) =
+            cx.with_new_id(|cx| crate::widget::Switch::new(cx.id_path(), self.is_on));
+        (id, (), element)
+    }
+
+    fn rebuild(
+        &self,
+        _cx: &mut Cx,
+        prev: &Self,
+        _id: &mut Id,
+        _state: &mut Self::State,
+        element: &mut Self::Element,
+    ) -> ChangeFlags {
+        if prev.is_on != self.is_on {
+            element.set_is_on(self.is_on)
+        } else {
+            ChangeFlags::default()
+        }
+    }
+
+    fn message(
+        &self,
+        _id_path: &[Id],
+        _state: &mut Self::State,
+        _message: Box<dyn Any>,
+        app_state: &mut bool,
+    ) -> MessageResult<A> {
+        *app_state = !*app_state;
+        MessageResult::Nop
```

Comment:
See above, I think this should invocate the callback instead of directly acting on the `app_state` (and then return a `MessageResult::Action`).

Reply:
I figured that the callback was necessary for the button as you don't know what the intention for it is. The function of the toggle switch though is much simpler, it just needs to reverse the previous state. In there a situation that a callback would be useful?

Reply:
Question: When you add the callback to the `MessageResult` does it get executed in `message()` and the resolved value is passed up the tree using the action or do you pass the callback. If the callback is the payload of the action who executes the callback?

Reply:
The resolved value will just be given in the callback which then actually modifies the app state (the button example could be seen as reference here).

"Optionally" the callback can return `A` (for Action). But I feel that this will likely be used rarely (i.e. `A` will be `()`) in the actual `app_logic` (IMHO more cognitive overhead than just a callback that modifies the app state).

It's executed by passing messages currently in the widget context(s) in e.g. `Widget::event` or `Widget::update`, that in turn propagate to the view via `View::message()` using the id path that then can mutate the app state via the callback.

There are other use cases though for a `MessageResult::Action`, one being async mutations outside of the immediate `app_logic` function (and something I'd like to continue to investigate in more detail, as I feel this will be quite relevant in bigger applications).

If you're interested, I have experimented a little bit here: https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/trui. Though I would approach this currently using the `MessageResult` directly (elm like enum, to be more composable, similar as in the todomvc example).